ext/contributioncancelactions/contributioncancelactions.php

```diff
@@ -0,0 +1,19 @@
+<?php
+
+require_once 'contributioncancelactions.civix.php';
+// phpcs:disable
+use CRM_Contributioncancelactions_ExtensionUtil as E;
+// phpcs:enable
+
+/**
+ * Implements hook_civicrm_preProcess().
+ *
+ * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_post
+ */
+function contributioncancelactions_civicrm_post($op, $objectName, $objectId, $objectRef) {
+  if ($op === 'edit' && $objectName === 'Contribution') {
+    if ($objectRef->contribution_status_id === CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Cancelled')) {
```

Comment:
@eileenmcnaughton This might be a bit of a pain to resolve but this doesn't actually do what we need - this would fire every time a contribution that has status "Cancelled" is edited. But what we actually need is a "contribution status was changed to cancelled" event.

Maybe we should consider introducing a set of symfony events/hooks that explicitly notify on eg. contribution cancel. I've implemented a few (custom) things like this and civirules does similar things but the only way to do it is rather messy because you have to capture data (ie. the before state) in the pre hook, store it statically and then compare it in the post hook.
Potentially we could find a place deep down in the BAO/DAO (eg. writeObjects) that would trigger "specialised" events like this. I'm not sure what's best, maybe @totten @colemanw have some ideas.

Reply:
@mattwire I was assuming we would largely get around that by checking for related pending entities & cancelling them - so the lack of pending entities would prevent it from acting twice